### ISSUE TITLE
PCA tool for IsoGlyP

### ISSUE THREAD

**Issue by Luke-Johnson-5** [glygener/glygen-issues#1562]
Give the script to run PCA analysis of IsoGlyP scoring.

**Comment by rykahsay** [comment]
Example input file (tmp/input.csv)
Site --> protein_ac|position combination
score --> score from isoglyp
dummy --> dummy field with all rows having 0.0 value for this field (needed for pca to work)
class --> 1 for known o-glyco sites, 0 for uknown o-glyco sites
```
site,score,dummy,class
P000001|100,8.0,0.0,1
P100002|200,9.0,0.0,1
P000003|300,8.5,0.0,1
P000004|450,7.0,0.0,1
P000005|150,2.0,0.0,0
P000006|250,3.0,0.0,0
P000007|350,4.0,0.0,0
P000008|400,5.0,0.0,0
```

<img width="549" alt="image" src="https://github.com/user-attachments/assets/4f4d5a85-b537-4498-83b5-41a6dcddda99">



Code
```
import matplotlib.pyplot as plt
import mpl_toolkits.mplot3d  # noqa: F401
from sklearn import datasets, decomposition

from sklearn.decomposition import PCA, IncrementalPCA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
import numpy


def load_features(in_file):

    X, y  = [], []
    with open(in_file, "r") as FR:
        lcount = 0
        for line in FR:
            lcount += 1
            if lcount == 1:
                continue
            row = line.split(",")
            X.append(row[1:-1])
            y.append(row[-1])
    X = numpy.array(X).astype(float)
    y = numpy.array(y).astype(int)
    return X, y


def main():

    in_file = "tmp/input.csv"
    X, y = load_features(in_file) 
    target_name_list = ["known", "unknown"]
    colors = ["green", "red"]

    pca = PCA(n_components=2)
    X_r = pca.fit(X).transform(X)
    plt.figure()
    lw = 2
    for color, i, target_name in zip(colors, [0, 1, 2], target_name_list):
        plt.scatter(
            X_r[y == i, 0], X_r[y == i, 1], color=color, alpha=0.8, lw=lw, label=target_name
        )
    plt.legend(loc="best", shadow=False, scatterpoints=1)
    plt.title("PCA of scores")
    #plt.show()
    plt.savefig('tmp/output.png') 
    return

if __name__ == '__main__':
    main()
```

**Comment by Luke-Johnson-5** [comment]
@rykahsay Yesterday you mentioned I could find the o-glycosylation information in an already existing dataset. I was looking through them but I am not 100% sure which I would be using for this. Which dataset should I be using for this?

**Comment by rykahsay** [comment]
Look at the commands given below: it looks like there are 33853 o-glycosites in version 2.5.1

```
$ cat /data/shared/glygen/releases/data/v-2.5.1/reviewed/*_proteoform_glycosylation_sites_* | grep ",\"O-linked\","  | awk -F, '{print $1,$2}' | grep -v "\"\"" |sort -u | wc
33853   67706  577263


$ cat /data/shared/glygen/releases/data/v-2.5.1/reviewed/*_proteoform_glycosylation_sites_* | grep ",\"O-linked\","  | awk -F, '{print $1,$2}' | grep -v "\"\"" |sort -u | head
"A0A087WXM9-1" "255"
"A0A087WXM9-1" "262"
"A0A087WXM9-1" "263"
"A0A0A6YVU8-1" "211"
"A0A0A6YVU8-1" "221"
"A0A0A6YW27-1" "223"
"A0A0A6YW27-1" "225"
"A0A0A6YXN8-1" "138"
"A0A0A6YXN8-1" "141"
"A0A0A6YXN8-1" "147"


```

**Comment by Luke-Johnson-5** [comment]
@rykahsay Had to mess with the script some to get it to run however here are the results for the PCA Script. Working on the plot of just the known site evals.
![image](https://github.com/user-attachments/assets/ee03903f-2fbb-4c03-8572-ceadd3544710)

**Comment by Luke-Johnson-5** [comment]
@rykahsay Here is the analysis of the known glycosylation site by score and cite. 
![image](https://github.com/user-attachments/assets/fc26f9f2-25e0-4879-bfe0-b3d89ba43099)

here is the raw excel files where I made this if you would like more information.
[PCA_testset_isoglyp.xlsx](https://github.com/user-attachments/files/16273400/PCA_testset_isoglyp.xlsx)

**Comment by rykahsay** [comment]
@Luke-Johnson-5  ... on the PCA plot, please tell me how you generated the input file for the PCA and share the input file here

**Comment by Luke-Johnson-5** [comment]
@rykahsay 
No problem. I made the test set using 5 random human proteins. These being the proteins seen above listed in the score distribution. I then ran this through the script /software/o_glycos/pca_make_input.py. This generated the attached file. This file was then ran through the PCA script you provided above.
[pca_ready.csv](https://github.com/user-attachments/files/16301819/pca_ready.csv)

**Comment by rykahsay** [comment]
@Luke-Johnson-5  ... why 5? I want you to do for all of them

**Comment by Luke-Johnson-5** [comment]
@rykahsay 
So I have gotten a PCA done for all of the human proteins and this what I was seeing. I was not aware the scores could go as high as they are here. However, After reevaluating the ways the scores are calculated It appears there is not a hard upper limit on the these scores. Here is what is specifically said on the IsoGlyP site about how the scores are to be understood. 

"An EVP value of 1 would indicate the transferase perceives the sequence as relatively neutral, i.e not inhibited or not enhanced, but nevertheless likely to be glycosylated. An EVP value greater than 1 would suggest a higher rate or likelihood of glycosylation, therefore a value of 2 would suggest a 2 fold rate of glycosylation. An EVP value of 0.2 would suggest a very poor site not likely to be glycosylated"

![image](https://github.com/user-attachments/assets/1b658bef-9b1e-4f22-8c7c-c3a538988af2)

**Comment by Luke-Johnson-5** [comment]
@rykahsay 
Here is something else very interesting to note that I found when looking through some of the data. I found this:
A known cite that is receiving a score of under 1 while an unknown cite receiving a score as high as 50.
![image](https://github.com/user-attachments/assets/51c76152-0c94-48f2-8190-bdc1a23272ff)

**Comment by rykahsay** [comment]
Next, please answer the following very carefully for each organism separately:

1. how many "known" sites do you have?
2. how many predicted sites do do you have in total?
3. how many predicted "known" sites do you have?
4. how many predicted "unknown" sites do you have?
5. how many predicted "uknown" sites are scoring above 25
6. how many predicted "uknown" sites are scoring above 30
7. how many predicted "uknown" sites are scoring above 35

Best way is to post a table with the following columns

organism,n_known_sites,n_predicted_sites_total,n_predicted_sites_known,n_predicted_sites_unknown,n_predicted_sites_unknown_gt_25,n_predicted_sites_unknown_gt_30,n_predicted_sites_unknown_gt_35

**Comment by Luke-Johnson-5** [comment]
@rykahsay 

| Organism    | n_known_sites | n_predicted_sites_total | n_predicted_sites_known | n_predicted_sites_unknown | n_predicted_sites_unknown_gt_25 | n_predicted_sites_unknown_gt_30 | n_predicted_sites_unknown_gt_35 |
| ----------- | ------------- | ----------------------- | ----------------------- | ------------------------- | ------------------------------- | ------------------------------- | ------------------------------- |
| human       | 22296         | 1567727                 | 22186                   | 1545541                   | 288                             | 121                             | 48                              |
| mouse       | 7108          | 1643489                 | 7078                    | 1636411                   | 210                             | 66                              | 34                              |
| pig         | 179           | 1572037                 | 171                     | 1571866                   | 269                             | 103                             | 67                              |
| rat         | 994           | 1618953                 | 974                     | 1617979                   | 232                             | 77                              | 44                              |
| sarscov1    | 4             | 2010                    | 3                       | 2007                      | 0                               | 0                               | 0                               |
| sarscov2    | 29            | 2055                    | 28                      | 2027                      | 0                               | 0                               | 0                               |
| yeast       | 2953          | 437794                  | 2947                    | 434847                    | 17                              | 6                               | 0                               |
| chicken     | 63            | 1471108                 | 55                      | 1471053                   | 170                             | 52                              | 19                              |
| dicty       | 16            | 1073418                 | 15                      | 1073403                   | 82                              | 14                              | 5                               |
| fruitfly    | 205           | 1040991                 | 204                     | 1040787                   | 129                             | 35                              | 12                              |
| arabidopsis | 329           | 1577239                 | 326                     | 1576913                   | 95                              | 22                              | 5                               |
| hcv1a       | 0             | 435                     | 0                       | 435                       | 0                               | 0                               | 0                               |
| hcv1b       | 6             | 445                     | 5                       | 440                       | 1                               | 1                               | 0                               |

[Isoglyp PCA data.xlsx](https://github.com/user-attachments/files/16444653/Isoglyp.PCA.data.xlsx)

**Comment by rykahsay** [comment]
@Luke-Johnson-5 .... please add the following columns and post updated table:


n_predicted_sites_unknown_gt_10 
n_predicted_sites_unknown_gt_15 
n_predicted_sites_unknown_gt_20

n_predicted_sites_known_gt_10 
n_predicted_sites_known_gt_15 
n_predicted_sites_known_gt_20
n_predicted_sites_known_gt_25 
n_predicted_sites_known_gt_30 
n_predicted_sites_known_gt_35

**Comment by Luke-Johnson-5** [comment]
@rykahsay 
Here is the updated table
| Organism    | n_known_sites | n_predicted_sites_total | n_predicted_sites_known | n_predicted_sites_unknown | n_predicted_sites_unknown_gt_25 | n_predicted_sites_unknown_gt_30 | n_predicted_sites_unknown_gt_35 | n_predicted_sites_unknown_gt_10 | n_predicted_sites_unknown_gt_15 | n_predicted_sites_unknown_gt_20 | n_predicted_sites_known_gt_10 | n_predicted_sites_known_gt_15 | n_predicted_sites_known_gt_20 | n_predicted_sites_known_gt_25 | n_predicted_sites_known_gt_30 | n_predicted_sites_known_gt_35 |
| ----------- | ------------- | ----------------------- | ----------------------- | ------------------------- | ------------------------------- | ------------------------------- | ------------------------------- | ------------------------------- | ------------------------------- | ------------------------------- | ----------------------------- | ----------------------------- | ----------------------------- | ----------------------------- | ----------------------------- | ----------------------------- |
| human       | 22296         | 1567727                 | 22186                   | 1545541                   | 288                             | 121                             | 48                              | 6806                            | 2120                            | 698                             | 280                           | 85                            | 33                            | 19                            | 7                             | 2                             |
| mouse       | 7108          | 1643489                 | 7078                    | 1636411                   | 210                             | 66                              | 34                              | 6361                            | 1837                            | 577                             | 60                            | 18                            | 9                             | 6                             | 1                             | 0                             |
| pig         | 179           | 1572037                 | 171                     | 1571866                   | 269                             | 103                             | 67                              | 6654                            | 1892                            | 663                             | 4                             | 0                             | 0                             | 0                             | 0                             | 0                             |
| rat         | 994           | 1618953                 | 974                     | 1617979                   | 232                             | 77                              | 44                              | 6350                            | 1750                            | 592                             | 9                             | 2                             | 0                             | 0                             | 0                             | 0                             |
| sarscov1    | 4             | 2010                    | 3                       | 2007                      | 0                               | 0                               | 0                               | 2                               | 2                               | 0                               | 0                             | 0                             | 0                             | 0                             | 0                             | 0                             |
| sarscov2    | 29            | 2055                    | 28                      | 2027                      | 0                               | 0                               | 0                               | 0                               | 0                               | 0                               | 0                             | 0                             | 0                             | 0                             | 0                             | 0                             |
| yeast       | 2953          | 437794                  | 2947                    | 434847                    | 17                              | 6                               | 0                               | 874                             | 171                             | 64                              | 2                             | 0                             | 0                             | 0                             | 0                             | 0                             |
| chicken     | 63            | 1471108                 | 55                      | 1471053                   | 170                             | 52                              | 19                              | 5570                            | 1512                            | 502                             | 1                             | 0                             | 0                             | 0                             | 0                             | 0                             |
| dicty       | 16            | 1073418                 | 15                      | 1073403                   | 82                              | 14                              | 5                               | 2977                            | 1022                            | 278                             | 3                             | 2                             | 0                             | 0                             | 0                             | 0                             |
| fruitfly    | 205           | 1040991                 | 204                     | 1040787                   | 129                             | 35                              | 12                              | 4375                            | 1176                            | 367                             | 1                             | 1                             | 1                             | 1                             | 0                             | 0                             |
| arabidopsis | 329           | 1577239                 | 326                     | 1576913                   | 95                              | 22                              | 5                               | 3465                            | 810                             | 277                             | 1                             | 1                             | 0                             | 0                             | 0                             | 0                             |
| hcv1a       | 0             | 435                     | 0                       | 435                       | 0                               | 0                               | 0                               | 3                               | 2                               | 0                               | 0                             | 0                             | 0                             | 0                             | 0                             | 0                             |
| hcv1b       | 6             | 445                     | 5                       | 440                       | 1                               | 1                               | 0                               | 2                               | 2                               | 1                               | 0                             | 0                             | 0                             | 0                             | 0                             | 0                             |
